File: sdk/python/aegis/receipt.py
from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from .verifier import ReceiptVerification, ReceiptVerifier


@dataclass(slots=True)
class Receipt:
    path: str
    raw: dict[str, Any]
    public_key_path: str | None = None
    summary_path: str | None = None
    verifier: ReceiptVerifier | None = field(default=None, repr=False)

    @classmethod
    def load(
        cls,
        path: str,
        *,
        public_key_path: str | None = None,
        summary_path: str | None = None,
        verifier: ReceiptVerifier | None = None,
    ) -> Receipt:
        raw = json.loads(Path(path).read_text(encoding='utf-8'))
        if public_key_path is None:
            inferred = Path(path).with_name('receipt.pub')
            if inferred.exists():
                public_key_path = str(inferred)
        if summary_path is None:
            inferred_summary = Path(path).with_name('receipt.summary.txt')
            if inferred_summary.exists():
                summary_path = str(inferred_summary)
        return cls(path=path, raw=raw, public_key_path=public_key_path, summary_path=summary_path, verifier=verifier)

    @property
    def proof_dir(self) -> str:
        return str(Path(self.path).parent)

    @property
    def statement(self) -> dict[str, Any]:
        return dict(self.raw.get('statement', {}))

    @property
    def predicate(self) -> dict[str, Any]:
        return dict(self.statement.get('predicate', {}))

    @property
    def execution_id(self) -> str | None:
        return self.predicate.get('execution_id')

    @property
    def verdict(self) -> str | None:
        divergence = self.predicate.get('divergence') or {}
        if isinstance(divergence, dict):
            return divergence.get('verdict')
        return None

    @property
    def signing_mode(self) -> str | None:
        trust = self.predicate.get('trust') or {}
        if isinstance(trust, dict):
            return trust.get('signing_mode')
        return None

    @property
    def key_source(self) -> str | None:
        trust = self.predicate.get('trust') or {}
        if isinstance(trust, dict):
            return trust.get('key_source')
        return None

    @property
    def summary_text(self) -> str | None:
        if not self.summary_path:
            return None
        summary_file = Path(self.summary_path)
        if not summary_file.exists():
            return None
        return summary_file.read_text(encoding='utf-8')

    def verify(self) -> ReceiptVerification:
        verifier = self.verifier or ReceiptVerifier()
        return verifier.verify_receipt(receipt_path=self.path, public_key_path=self.public_key_path)

File: sdk/typescript/src/receipt.ts
import { existsSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";

import { ReceiptVerification, ReceiptVerifier } from "./verifier.js";

export class Receipt {
  readonly path: string;
  readonly raw: Record<string, unknown>;
  readonly publicKeyPath?: string;
  readonly summaryPath?: string;
  private readonly verifier?: ReceiptVerifier;

  constructor(input: { path: string; raw: Record<string, unknown>; publicKeyPath?: string; summaryPath?: string; verifier?: ReceiptVerifier }) {
    this.path = input.path;
    this.raw = input.raw;
    this.publicKeyPath = input.publicKeyPath;
    this.summaryPath = input.summaryPath;
    this.verifier = input.verifier;
  }

  static load(path: string, options: { publicKeyPath?: string; summaryPath?: string; verifier?: ReceiptVerifier } = {}): Receipt {
    const raw = JSON.parse(readFileSync(path, "utf8")) as Record<string, unknown>;
    let publicKeyPath = options.publicKeyPath;
    let summaryPath = options.summaryPath;
    if (!publicKeyPath) {
      const inferred = join(dirname(path), "receipt.pub");
      if (existsSync(inferred)) publicKeyPath = inferred;
    }
    if (!summaryPath) {
      const inferred = join(dirname(path), "receipt.summary.txt");
      if (existsSync(inferred)) summaryPath = inferred;
    }
    return new Receipt({ path, raw, publicKeyPath, summaryPath, verifier: options.verifier });
  }

  get proofDir(): string {
    return dirname(this.path);
  }

  get statement(): Record<string, unknown> {
    const statement = this.raw.statement;
    return statement && typeof statement === "object" ? { ...(statement as Record<string, unknown>) } : {};
  }

  get predicate(): Record<string, unknown> {
    const predicate = this.statement.predicate;
    return predicate && typeof predicate === "object" ? { ...(predicate as Record<string, unknown>) } : {};
  }

  get executionId(): string | undefined {
    return typeof this.predicate.execution_id === "string" ? this.predicate.execution_id : undefined;
  }

  get resultClass(): string | undefined {
    return typeof this.predicate.result_class === "string" ? this.predicate.result_class : undefined;
  }

  get divergenceVerdict(): string | undefined {
    const divergence = this.predicate.divergence;
    if (divergence && typeof divergence === "object" && typeof (divergence as Record<string, unknown>).verdict === "string") {
      return (divergence as Record<string, unknown>).verdict as string;
    }
    return undefined;
  }

  get signingMode(): string | undefined {
    const trust = this.predicate.trust;
    if (trust && typeof trust === "object" && typeof (trust as Record<string, unknown>).signing_mode === "string") {
      return (trust as Record<string, unknown>).signing_mode as string;
    }
    return undefined;
  }

  get keySource(): string | undefined {
    const trust = this.predicate.trust;
    if (trust && typeof trust === "object" && typeof (trust as Record<string, unknown>).key_source === "string") {
      return (trust as Record<string, unknown>).key_source as string;
    }
    return undefined;
  }

  get summaryText(): string | undefined {
    if (!this.summaryPath || !existsSync(this.summaryPath)) return undefined;
    return readFileSync(this.summaryPath, "utf8");
  }

  async verify(): Promise<ReceiptVerification> {
    const verifier = this.verifier ?? new ReceiptVerifier();
    return verifier.verifyReceipt({ receiptPath: this.path, publicKeyPath: this.publicKeyPath });
  }
}
